### ISSUE TITLE
Add CLI subcommands to the RuntimeBuilder API

### ISSUE THREAD

**Issue by marceline-cramer** [hearth-rs/hearth#68]
Currently, `hearth-ctl` commands are built into the binary and plugin-dependent subcommands (like #52) have to be specialized for each plugin's APIs. This decoupling effectively makes `hearth-ctl` an OS-available proxy to the process environment, but it only configures plugins via existing message protocols that would be available to any other process. It also makes developing configuration for plugins more difficult, because we need to separately develop a protocol for messages sent to plugins' services and `ctl`'s subcommand front-ends for those messages.

Alternatively, we could add support for CLI commands as part of a plugin's host-side API. The runtime would contain a registry of each plugin's commands and `hearth-ctl` would dynamically query and execute those commands over IPC. This provides an alternative, trusted method of configuring plugins from the host that eliminates friction between low-level message protocols and user-friendly, shell-executable commands. This also improves Hearth's runtime flexibility because we can rapidly develop ways to modify the behavior of plugins without needing to update a configuration file or restart the runtime.

Broad overview of steps involved:
- add a daemon-only RPC interface for querying and executing commands
- add a method on `RuntimeBuilder` to add named commands to the runtime
- add subcommands to `hearth-ctl` to display query results for commands
- add support for execution of plugin commands to the `hearth-ctl` CLI

Notes:
- The command data structure added to runtimes during building can be a struct containing a function pointer and metadata like its name. The plugin does not need to do any manual event/channel handling for the calling of commands, and plugins can manage their own command-related event loops and state.
- The clap crate can (and will probably frequently) be used for easily parsing plugin commands, although it's not mandatory.
- The format of the command execution IPC protocol should be an `argv`-style list of strings, a list of environment variables, and a working directory path in, and the command should return an output string and the return code. Interactive commands will not be supported; plugin commands are one-and-done for the sake of simplicity.
- The working directory and environment variables of `ctl` will need to be transferred over IPC to plugins. One usecase of this includes loading WebAssembly modules from the file system in `spawn-wasm`.

Subjects of questioning/discussion:
- [ ] How do we handle naming collisions in commands i.e. if commands in two different plugins, which may not be written by coordinated individuals, accidentally end up having the same name? How does the runtime identify which commands come from which plugins?
- [ ] What does the UX look like with `hearth-ctl`'s CLI for calling plugin commands?
- [ ] How are  conflicts between environment variables from the runtime's environment and `hearth-ctl`'s environment resolved?
- [ ] How are `stdout` and `stderr` handled in returned output data? Are they both included? If so, how do we encode them in the protocol? Is ordering important?

**Comment by airidaceae** [comment]
I think that collisions in commands could be best handled by having both a shorthand name and a full name for the subcommand. For example if a a plugin named `wasm` had a subcommand called `spawn` the command could be referenced by `hearth-ctl spawn` unless there is a conflict with either the core commands or a command added by another plugin. In that case it would require that the subcommand is referenced by its full name, in this case `hearth-ctl wasm.spawn`. 

Calling a subcommand by its full name would be best used in scripts where it is important for the command to not fail randomly because of another installed crate. The shorthand could be best utilized when repeatedly typing out many commands by hand especially for plugins that are core adjacent. I think that this concept could also be extended to the core commands themselves with something like `hearth-ctl core.list-peers`.

**Comment by airidaceae** [comment]
I believe this can be closed now as @marceline-cramer and I decided that this would not be a feature for `hearth-ctl` itself but would instead be another system. We should make an issue for what we discussed and then close this issue.

**Comment by marceline-cramer** [comment]
Yeah, I think that the general gist is that we'd make a *guest-side* store of commands accessible via a service where different services can register service-specific commands to that store. Then, we'd write a separate CLI tool that queries that store service and can invoke commands from the command-line.

**Comment by airidaceae** [comment]
This is an issue that should be on [kindling](https://github.com/hearth-rs/kindling)

**Comment by airidaceae** [comment]
kindling now exists on the main repo so it belongs here again!

**Comment by marceline-cramer** [comment]
Could we actually open a new issue? The idea has Ship of Theseus'd quite a bit aways from the original issue.
